### PR TITLE
Feat/depin mining weight

### PR DESCRIPTION
**Pull request by ludete**

### CODE REVIEW

File: contracts/gauges/Gauge.sol [hunk 1]
```diff
@@ -53,7 +69,33 @@ contract Gauge is IGauge, ERC2771Context, ReentrancyGuard {
...
+    shares[_user] = _share;
+    totalShare = totalShare - _oldShare + _share;
+    updateGainBalance(_user);
+  }
+
+  function updateGainBalance(address _user) internal {
+    uint256 _originBalance = balanceOf[_user];
+    if ( _originBalance > 0 ){
+      //
+      uint256 _gainBalance = _originBalance * gain / Factor + _originBalance;
```
Comment (ludete) [inline contracts/gauges/Gauge.sol:94]: _originBalance * gain * share / Factor + _originBalance;

File: contracts/gauges/Gauge.sol [hunk 2]
```diff
@@ -53,7 +69,34 @@ contract Gauge is IGauge, ERC2771Context, ReentrancyGuard {
...
+  }
+
+  function updateGainBalance(address _user) internal {
+    uint256 _originBalance = balanceOf[_user];
+    if ( _originBalance > 0 ){
+      uint256 _share = shares[_user];
+      uint256 _gainBalance = _originBalance * gain / Factor * _share / gainTotalSupply + _originBalance;
+      uint256 _oldGainBalance = gainBalanceOf[_user];
+      gainBalanceOf[_user] = _gainBalance;
+      gainTotalSupply = gainTotalSupply - _oldGainBalance + _gainBalance;
```
Comment (CoderZhi) [inline contracts/gauges/Gauge.sol:98]:
```
if totalShare > 0 {
  gainTotalSupply = totalSupply * (1 + Factor);
} else {
  totalGainSupply = totalSupply;
}
```
such that, `gainTotalSupply` can be deleted.

File: contracts/gauges/Gauge.sol [hunk 3]
```diff
@@ -53,7 +69,34 @@ contract Gauge is IGauge, ERC2771Context, ReentrancyGuard {
...
+    updateGainBalance(_user);
+  }
+
+  function updateGainBalance(address _user) internal {
+    uint256 _originBalance = balanceOf[_user];
+    if ( _originBalance > 0 ){
+      uint256 _share = shares[_user];
+      uint256 _gainBalance = _originBalance * gain / Factor * _share / gainTotalSupply + _originBalance;
+      uint256 _oldGainBalance = gainBalanceOf[_user];
+      gainBalanceOf[_user] = _gainBalance;
```
Comment (CoderZhi) [inline contracts/gauges/Gauge.sol:97]: `gainBalance = balance * (Factor * share / totalShare + 1)`
such that, `gainBalanceOf` can be deleted.

File: contracts/gauges/Gauge.sol [hunk 4]
```diff
@@ -53,7 +69,34 @@ contract Gauge is IGauge, ERC2771Context, ReentrancyGuard {
...
+  }
+
+//  function rewardPerTokenWithShare(address _user) public view returns (uint256)  {
+//    uint256 _pertoken = rewardPerToken();
+//    uint256 _reward = _pertoken * wLP / Factor + _pertoken * wVote / Factor;
+////    _pertoken *(wlp/factor + wVote/factor)
+//    return _reward;
+//  }
+
+  function depositShare(address _user, uint256 _share) external {
```
[the remark this answers is not in the document]
Comment (ludete) [reply contracts/gauges/Gauge.sol:82]: ok

File: contracts/gauges/Gauge.sol (hunk 2 above)
Comment (ludete) [reply contracts/gauges/Gauge.sol:98]: 保留这个状态比较好，rewardPerToken() 计算时会使用，否则每次都要计算，更耗gas

File: contracts/gauges/Gauge.sol (hunk 3 above)
Comment (ludete) [reply contracts/gauges/Gauge.sol:97]: same reason with above

File: contracts/gauges/Gauge.sol (hunk 2 above)
Comment (CoderZhi) [reply contracts/gauges/Gauge.sol:91]:
```
function rewardPerToken() public view returns (uint256) {
    if (totalSupply == 0) {
        return rewardPerTokenStored;
    }
    uint256 t = lastTimeRewardApplicable() - lastUpdateTime;
    if (t == 0) {
        return rewardPerTokenStored;
    }
    if (totalShare == 0) {
        return rewardPerTokenStored + (t * rewardRate * PRECISION) / totalSupply;
    }
    return rewardPerTokenStored + (t * rewardRate * PRECISION) / totalSupply / (1 + Factor);
  }
```
most of the cases, `t == 0`. So, comparing to current implementation, we save storage gas as well as computing gas in func `updateGainBalance`

**Review by CoderZhi — APPROVED** [review]
The only concern is what type `shareFactor` could be, int > 0 or float > 0

File: contracts/gauges/Gauge.sol [hunk 5]
```diff
@@ -107,16 +105,20 @@ contract Gauge is IGauge, ERC2771Context, ReentrancyGuard {
...
   }
 
-  function gainBalance(address _account) public view returns (uint256) {
+  function updateGainBalance(address _account) internal {
     uint256 _share = shares[_account];
     uint256 _originBalance = balanceOf[_account];
+    uint256 _gainBalance = _originBalance * TOKENLESS_PRODUCTION / BASE;
     if ( _originBalance > 0 && _share > 0 ){
-      return _originBalance * (shareFactor * _share / totalShare + 1);
+      _gainBalance += totalSupply * _share / totalShare *(BASE - TOKENLESS_PRODUCTION) / BASE;
```
Comment (CoderZhi) [inline contracts/gauges/Gauge.sol]: `_gainBalance += totalSupply * _share * (BASE - TOKENLESS_PRODUCTION) / totalShare / BASE;`

File: contracts/gauges/Gauge.sol [hunk 6]
```diff
@@ -46,20 +48,18 @@ contract Gauge is IGauge, ERC2771Context, ReentrancyGuard {
   mapping(address => uint256) public shares;
   /// @inheritdoc IGauge
   uint256 public totalShare;
+
+  /// @inheritdoc IGauge
+  mapping(address => uint256) public gainBalanceOf;
   /// @inheritdoc IGauge
-  uint256 public shareFactor;
+  uint256 public totalGainBalance;
```
Comment (CoderZhi) [inline contracts/gauges/Gauge.sol]: `totalWeightedBalance`

File: contracts/gauges/Gauge.sol [hunk 7]
```diff
@@ -81,6 +78,7 @@ contract Gauge is IGauge, ERC2771Context, ReentrancyGuard {
     shares[_user] = _share;
     totalShare = totalShare - _oldShare + _share;
     _updateRewards(_user);
```
Comment (CoderZhi) [inline contracts/gauges/Gauge.sol]: move `_updateRewards` to line 77

**Review by CoderZhi — APPROVED** [review]

File: contracts/gauges/Gauge.sol (hunk 7 above)
Comment (ludete) [reply contracts/gauges/Gauge.sol]: _updateRewards only needs to be before updateGainBalance, which only use gainBalance & totalGainBalance

Comment (CoderZhi) [reply contracts/gauges/Gauge.sol]: moving to L77 will make the code style consistent: updateRewards before any other updates to the user/contract

Comment (ludete) [reply contracts/gauges/Gauge.sol]: agree.

**Pull request merged**